npm/private/npm_translate_lock_rule.bzl

```diff
@@ -0,0 +1,220 @@
+"Convert pnpm lock file into starlark Bazel fetches"
+
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load(":npm_translate_lock.bzl", "helpers")
+load(":npm_translate_lock_state.bzl", "npm_translate_lock_state")
+load(":utils.bzl", "utils")
+load(":transitive_closure.bzl", "translate_to_transitive_closure")
+
+RULES_JS_FROZEN_PNPM_LOCK_ENV = "ASPECT_RULES_JS_FROZEN_PNPM_LOCK"
+
+########################################################################################################################
+def _impl(rctx):
+    rctx.report_progress("Initializing")
+
+    state = npm_translate_lock_state.new(rctx)
+
+    # If a pnpm lock file has not been specified then we need to bootstrap by running `pnpm
+    # import` in the user's repository
+    if not rctx.attr.pnpm_lock:
+        _bootstrap_import(rctx, state)
+
+    if state.update_pnpm_lock():
+        # Run `pnpm install --lockfile-only` or `pnpm import` if its inputs have changed since last update
+        if state.action_cache_miss():
+            _fail_if_frozen_pnpm_lock(rctx, state)
+            if _update_pnpm_lock(rctx, state):
+                # If the pnpm lock file was changed then reload it before translation
+                state.reload_lockfile()
+
+    helpers.verify_node_modules_ignored(rctx, state.importers(), state.root_package())
+
+    rctx.report_progress("Translating {}".format(state.label_store.relative_path("pnpm_lock")))
+
+    importers, packages = translate_to_transitive_closure(
+        state.importers(),
+        state.packages(),
+        rctx.attr.prod,
+        rctx.attr.dev,
+        rctx.attr.no_optional,
+    )
+
+    rctx.report_progress("Generating starlark for npm dependencies")
+
+    helpers.generate_repository_files(
+        rctx,
+        state.label_store.label("pnpm_lock"),
+        importers,
+        packages,
+        state.root_package(),
+        state.default_registry(),
+        state.npm_registries(),
+        state.npm_tokens(),
+        state.npm_basic_auth(),
+        state.link_workspace(),
+    )
+
+npm_translate_lock = repository_rule(
```

Comment:
nit: I'd wish for this file to be `npm_translate_lock.bzl` since it declares the rule. What could the other one be named?
oh I see you already have a TODO for that after landing.